python/util.py

```diff
@@ -84,3 +86,335 @@ def constant(
+        tgt = tgt.sb.col, tgt.sb.row
+
+        for n in DG.nodes:
+            if n in {src, tgt}:
+                continue
+
+            # what goes in must come out
+            model.Add(
+                sum([flow_var[e] for e in DG.in_edges(nbunch=n)])
+                == sum([flow_var[e] for e in DG.out_edges(nbunch=n)])
+            )
+
+            # flow must leave src, and must not enter src
+            model.Add(sum([flow_var[src, j] for j in DG.neighbors(src)]) == 1)
+            model.Add(sum([flow_var[i, src] for i in DG.neighbors(src)]) == 0)
+
+            # flow must enter tgt, and must not leave tgt
+            model.Add(sum([flow_var[tgt, j] for j in DG.neighbors(tgt)]) == 0)
+            model.Add(sum([flow_var[i, tgt] for i in DG.neighbors(tgt)]) == 1)
+
+    # Create demand variables
+    total_demand = {
+        (i, j): model.NewIntVar(0, len(flat_flow_vars), "") for i, j in DG.edges
+    }
+    overlapping_demands = {
+        (i, j): model.NewIntVar(0, len(flat_flow_vars), "") for i, j in DG.edges
+    }
+    used_edges = {(i, j): model.NewIntVar(0, 1, "") for i, j in DG.edges}
+
+    # Add demand/flow relationship
+    for i, j, attrs in DG.edges(data=True):
+        model.Add(total_demand[i, j] == sum([f[i, j] for f in flat_flow_vars]))
+        model.Add(total_demand[i, j] <= attrs["capacity"])
+
+        if min_edges:
+            model.AddMaxEquality(used_edges[i, j], [f[i, j] for f in flat_flow_vars])
+        else:
+            overlapping_flows = {}
+            for k, f1 in enumerate(flat_flow_vars):
+                for l, f2 in enumerate(flat_flow_vars[k + 1 :], start=k + 1):
+                    overlapping_flows[k, l] = model.NewIntVar(
+                        0, len(flat_flow_vars), ""
+                    )
+                    model.AddMultiplicationEquality(
+                        overlapping_flows[k, l], [f1[i, j], f2[i, j]]
+                    )
+
+            model.Add(
+                overlapping_demands[i, j]
+                == sum(
+                    [
+                        overlapping_flows[k, l]
+                        for k, f1 in enumerate(flat_flow_vars)
+                        for l, f2 in enumerate(flat_flow_vars[k + 1 :], start=k + 1)
+                    ]
+                )
+            )
+
+    obj = sum([total_demand[i, j] for i, j in DG.edges])
+    if min_edges:
+        obj += sum([used_edges[i, j] for i, j in DG.edges])
+    else:
+        obj += sum([overlapping_demands[i, j] for i, j in DG.edges])
+    model.Minimize(obj)
+
+    status = solver.Solve(model)
+    if status == cp_model.OPTIMAL or status == cp_model.FEASIBLE:
+        flow_paths = {}
+        for flow, flow_varss in flow_vars.items():
+            flow_paths[flow] = [
+                (i, j) for i, j in DG.edges if solver.Value(flow_varss[i, j]) > 0.5
+            ]
+
+        return flow_paths
+
+
+def route_using_ilp(DG, flows):
+    import gurobipy as gp
+    from gurobipy import GRB
```

Comment:
I was wondering how it was possible to have a solver used in the neighborhood without notice...
https://peps.python.org/pep-0008/#imports says:
> Imports are always put at the top of the file, just after any module comments and docstrings, and before module globals and constants.

Any serious reason to deviate?
And actually it looks like there are a lot of other `import` all over the place.

Reply:
Importing within function scope is very common exactly for this purpose - you do not want to trigger an `ImportError` for this file (generically called `util.py`) just because Gurobi isn't installed if the user never tries to run any of the ILP related functions.

Reply:
That sounds like a compelling use case.
Why not adding a comment about this?